### PR TITLE
fix: handle mandatory shipToGSTIN for e-invoice

### LINKED ISSUE #4336: e-Waybill Changes from 15 June(now on hold).
https://tutorial.gst.gov.in/downloads/news/gstn_advisory_ewb_approved.pdf

### PR DESCRIPTION
**Pull request by Anish59312**
closes: #4336

- [x] manual test for e way bill creation without IRN
- [ ] changes required to allow SEZ e-waybill and e-invoice sandbox mode

# Mandatory Ship-To GSTIN (GSTN advisory 17.06.2026, effective 01.08.2026)

Implements the Ship-To GSTIN changes for the e-Invoice and e-Waybill APIs, and fixes NIC
error **618** reported against `61d0027db` (PR #4342), where two addresses of the same party
were sent as a Bill To - Ship To transaction.

## Problem

`transaction_type` was derived purely from the **address link** being different:

```python
has_different_to_address = address.ship_to and address.ship_to != address.bill_to
```

So a customer's HQ (billing) and warehouse (shipping) — two Address records sharing one GSTIN —
produced `transactionType: 2` with `shipToGSTIN == toGstin`, which NIC rejects:

> **618**: Bill-to and ship-to gstin should not be same for this transaction type

The same shape fails the e-Invoice API with **2323**. Both are unreachable today only because
`shipToGSTIN` isn't sent in production until 01.08.2026 — it is already reproducible in sandbox.

### 1. Distinct party characterisation — `utils/e_waybill.py`, `utils/e_invoice.py`

A Bill To - Ship To transaction requires the consignee to be a party distinct from the buyer.
Two addresses of the same party are a Regular transaction. `URP` denotes a missing GSTIN rather
than an identity, so an unregistered consignee stays distinct from an unregistered buyer.

- **e-Waybill**: `transaction_type` degrades `2 -> 1` / `4 -> 3`; `shipToGSTIN` / `shipToTradeName`
  are not sent. The destination is unaffected — `toAddr1`, `toPincode` and `actToStateCode` are
  read from `ship_to` regardless of transaction type.
- **e-Invoice**: `ShipDtls` is omitted.

### 2. Shipping address is locked once reported — `overrides/transaction.py`

Ship To details reported in the e-Invoice can't be replaced afterwards (**2324**), and the shipping
address is the e-Waybill destination, so a change to `shipping_address_name` now reaches the
existing after-submit check. Sales Invoice only, as that's where the IRN lives.

### 3. Sandbox Ship To party — `constants/e_waybill.py`

`SANDBOX_SHIP_TO` gives the consignee a GSTIN of its own, as it can't be the same as bill to.
State and pincode are part of it, as the e-Invoice APIs validate them against the GSTIN
(**2325**, **3039**).

## Not included: `ExpShipDtls`

The advisory allows Ship To details to be supplied at e-Waybill generation (§4, §7B) where the
e-Invoice was generated without them. **This PR deliberately does not implement that.** The
shipping address is simply restricted instead. When the address genuinely needs to change, the
intended direction is to generate the e-Waybill directly using the IRN — a separate feature.

Consequence: for a same-GSTIN consignee, the shipping address is reported in neither document.
2323 / 4073 fire on GSTIN equality alone, so there is no payload that reports it with a truthful
GSTIN. This is a gap in the GSTN model, worth raising with them.

### Live NIC sandbox

Exercised end to end through `generate_e_invoice()` -> `_generate_e_waybill()`, with positive
**and** negative cases. Field names confirmed by acceptance, not assumption.

| § | Rule | Code | Result |
|---|---|---|---|
| 3, 7A | `ShipDtls.Gstin` mandatory where ship details provided | 5002 | accepted |
| 3, 6 | `URP` where GSTIN unavailable | — | accepted |
| 5, 7A | Bill To GSTIN != Ship To GSTIN | 2323 | rejected when forced equal |
| 7A | Ship To state matches GSTIN state | 2325 | rejected on mismatch |
| 7A | Ship To PIN belongs to state | 3039 | rejected when out of state |
| 9 | Old same-GSTIN IRNs generate a regular e-Waybill | 618 | type 1 |

Also verified: the same-GSTIN restriction applies at the by-IRN stage too (**4073**); an SEZ GSTIN is
accepted as Ship To GSTIN; sales returns characterise as type 3 with no `shipToGSTIN`; and an
e-Waybill generated without IRN is unaffected by ship details already reported in the e-Invoice.

### Tests

All driving real documents through the production entry points rather than planted state —
`generate_e_invoice()` with the mocked IRP response, and `mark_e_waybill_as_generated()`.

- distinct party: type 2 / 4 preserved, same GSTIN degrades to type 1 / 3, `URP` stays distinct
- shipping address lock: after the e-Waybill, and after the e-Invoice alone
- rollout gating: `shipToGSTIN` / `shipToTradeName` absent in production before 01.08.2026,
  present on and after (both the API payload and the offline JSON)

## Notes for review

- **The shipping address lock is not date gated.** It applies as soon as this is deployed, to
  documents that already have an e-Waybill or IRN.
- No schema change and no patch, so nothing to migrate.
- Existing tests and fixtures are unchanged, except `data/test_e_invoice.json`, where
  `shipping_details` asserted `ShipDtls.Gstin == BuyerDtls.Gstin` — a payload NIC now rejects
  with 2323.
- **Not included**: Voluntary e-Waybill Closure (advisory §10–15) is a separate feature.

### CODE REVIEW

File: india_compliance/gst_india/utils/e_waybill.py [hunk 1]
```diff
@@ -1273,19 +1273,30 @@ def get_data_with_irn(self):
...
+            "TransMode": str(self.transaction_details.mode_of_transport),
+            "TransId": self.transaction_details.gst_transporter_id,
+            "TransName": self.transaction_details.transporter_name,
+            "TransDocDt": self.transaction_details.lr_date,
+            "TransDocNo": self.transaction_details.lr_no,
+            "VehNo": self.transaction_details.vehicle_no,
+            "VehType": self.transaction_details.vehicle_type,
+        }
+
+        if 1 or is_e_waybill_changes_applicable(self.settings):
```
Comment (ljain112) [inline india_compliance/gst_india/utils/e_waybill.py]: .

File: india_compliance/gst_india/utils/test_e_invoice.py [hunk 2]
```diff
@@ -104,6 +157,55 @@ def test_request_data_for_foreign_transactions(self):
...
+            is_in_state=True,
+            do_not_submit=True,
+        )
+        si.submit()
+
+        invoice_data = EInvoiceData(si).get_data()
+        ship_dtls = invoice_data.get("ShipDtls")
+
+        self.assertIsNotNone(ship_dtls)
+        self.assertEqual(ship_dtls.get("Gstin"), "URP")
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/test_e_invoice.py]: this is fine to do

File: india_compliance/gst_india/utils/e_waybill.py [hunk 3]
```diff
@@ -1274,19 +1274,43 @@ def get_data_with_irn(self):
...
+                "TrdNm": self.ship_to.legal_name,
+                "Addr1": self.ship_to.address_line1,
+                "Addr2": self.ship_to.address_line2,
+                "Loc": self.ship_to.city,
+                "Pin": self.ship_to.pincode,
+                "Stcd": str(self.ship_to.state_number),
             }
-        )
+
+        return self.sanitize_data(data)
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_waybill.py]: transaction types 2 and 4 mean that the ship to address is available and not different than the bill to address. So this case is covered

File: india_compliance/gst_india/utils/test_e_waybill.py [hunk 4]
```diff
@@ -1431,32 +1433,83 @@ def test_ship_to_gstin_gated_by_rollout_date(self):
...
-        )
+            data = EWaybillData(si).get_data(with_irn=True)
+            # when with irn is true, this data is compatable with E-Invoice APIs
+            self.assertNotIn("ExpShipDtls", data)
+
+        # on/after rollout -> ExpShipDtls sent
+        with time_machine.travel(rollout_date, tick=False):
+            data = EWaybillData(si).get_data(with_irn=True)
+            self.assertIn("ExpShipDtls", data)
+            self.assertTrue(data["ExpShipDtls"].get("Gstin"))
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/test_e_waybill.py]: updated

**Comment by Anish59312** [comment]
follows https://einv-apisandbox.nic.in/release-notes.html

File: india_compliance/gst_india/utils/e_waybill.py [hunk 5]
```diff
@@ -1274,19 +1274,41 @@ def get_data_with_irn(self):
...
+
+    def irn_has_ship_to_details(self):
+        invoice_data = frappe.db.get_value("e-Invoice Log", self.doc.irn, "invoice_data")
+        if not invoice_data:
+            return ""
+        return (json.loads(invoice_data).get("ShipDtls") or {}).get("Gstin", "")
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_waybill.py]: resolved

File: india_compliance/patches/v16/set_is_generated_with_ship_to_in_e_invoice_log.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    # e-Invoice Log is named after the IRN
+    irns_generated_with_ship_to = (
+        frappe.qb.from_(si)
+        .select(si.irn)
+        .where(IfNull(si.irn, "") != "")
+        .where(IfNull(si.ewaybill, "") == "")
+        .where(IfNull(si.shipping_address_name, "") != "")
+        .where(si.shipping_address_name != IfNull(si.customer_address, ""))
+    )
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/patches/v16/set_is_generated_with_ship_to_in_e_invoice_log.py]: made the changes

File: india_compliance/gst_india/overrides/transaction.py [hunk 7]
```diff
@@ -1919,6 +1925,23 @@ def has_changed(field):
...
 
+def validate_shipping_address_change(doc):
+    if not doc.get("irn"):
+        return
+
+    # e-Invoice Log is named after the IRN
+    if not frappe.get_cached_value("e-Invoice Log", doc.irn, "is_generated_with_ship_to"):
+        return
+
+    if not doc.has_value_changed("shipping_address_name"):
```
Comment (ljain112) [inline india_compliance/gst_india/overrides/transaction.py]: this should come first ?

File: india_compliance/gst_india/utils/e_invoice.py [hunk 8]
```diff
@@ -830,8 +848,10 @@ def get_invoice_data(self):
                     "pincode": 500055,
                 }
                 self.billing_address.update(buyer)
-                if self.shipping_address:
-                    self.shipping_address.update(buyer)
+
+                # consignee is a different party here, so it needs a GSTIN of its own
+                if self.shipping_address and self.shipping_address.gstin != "URP":
```
Comment (ljain112) [inline india_compliance/gst_india/utils/e_invoice.py:845]: What if it is the same as the customer's GSTIN?

File: india_compliance/gst_india/utils/e_waybill.py [hunk 9]
```diff
@@ -1274,19 +1275,38 @@ def get_data_with_irn(self):
...
+            "VehType": self.transaction_details.vehicle_type,
+        }
+        if (
+            is_e_waybill_changes_applicable(self.settings)
+            and self.transaction_details.transaction_type in SHIP_TO_TRANSACTION_TYPES
+            and not frappe.get_cached_value(
+                "e-Invoice Log", {"irn": self.doc.irn, "is_generated_with_ship_to": 1}
+            )
+        ):
+            if self.sandbox_mode and self.ship_to.gstin and self.ship_to.gstin != "URP":
```
Comment (ljain112) [inline india_compliance/gst_india/utils/e_waybill.py]: can be a utility

File: india_compliance/gst_india/utils/e_waybill.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_waybill.py]: it has been removed

File: india_compliance/gst_india/utils/e_invoice.py [hunk 10]
```diff
@@ -381,6 +382,14 @@ def log_and_process_e_invoice_generation(doc, result, sandbox_mode=False, messag
         )
         invoice_data = frappe.as_json(decoded_invoice, indent=4)
 
+    # ship-to sent during IRN generation can be replaced at e-Waybill generation for
+    # exports, but not for B2B / SEZ (NIC error 2324)
+    is_generated_with_ship_to = (
+        doc.shipping_address_name
+        and doc.shipping_address_name != doc.customer_address
+        and doc.gst_category in ("Registered Regular", "SEZ")
+    )
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_invoice.py]: change is made

File: india_compliance/gst_india/overrides/transaction.py (hunk 7 above)
Comment (Anish59312) [reply india_compliance/gst_india/overrides/transaction.py]: yes

File: india_compliance/patches/v16/set_is_generated_with_ship_to_in_e_invoice_log.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    # e-Invoice Log is named after the IRN
+    irns_generated_with_ship_to = (
+        frappe.qb.from_(si)
+        .select(si.irn)
+        .where(IfNull(si.irn, "") != "")
+        .where(IfNull(si.ewaybill, "") == "")
+        .where(IfNull(si.shipping_address_name, "") != "")
+        .where(si.shipping_address_name != IfNull(si.customer_address, ""))
+        .where(si.gst_category.isin(("Registered Regular", "SEZ")))
```
[the remark this answers is not in the document]
Comment (Anish59312) [reply india_compliance/patches/v16/set_is_generated_with_ship_to_in_e_invoice_log.py]: changes are made in is_generated_with_ship_to to not be true when billing and shipping gstin are same

File: india_compliance/gst_india/utils/e_invoice.py (hunk 8 above)
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_invoice.py:845]: get_shipping_address will already return none setting the shipping_address to none for that case 
i have written tests for these edge cases (running outside sandbox mode and checking the get_data 's values)

File: india_compliance/gst_india/utils/e_waybill.py (hunk 9 above)
Comment (Anish59312) [reply india_compliance/gst_india/utils/e_waybill.py]: for now have left it as is

File: india_compliance/gst_india/overrides/transaction.py [hunk 12]
```diff
@@ -1919,6 +1925,23 @@ def has_changed(field):
...
 
 
+def validate_shipping_address_change(doc):
+    if not doc.get("irn"):
+        return
+
+    if not doc.has_value_changed("shipping_address_name"):
+        return
+
+    # e-Invoice Log is named after the IRN
```
Comment (ljain112) [inline india_compliance/gst_india/overrides/transaction.py]: should we check gst category ?

File: india_compliance/gst_india/utils/e_invoice.py [hunk 13]
```diff
@@ -381,6 +382,16 @@ def log_and_process_e_invoice_generation(doc, result, sandbox_mode=False, messag
...
 
+    # ship-to sent during IRN generation can be replaced at e-Waybill generation for
+    # exports, but not for B2B / SEZ (NIC error 2324)
+    is_generated_with_ship_to = bool(
+        doc.shipping_address_name
+        and doc.shipping_address_name != doc.customer_address
+        and doc.gst_category in ("Registered Regular", "SEZ")
+        and frappe.get_cached_value("Address", doc.shipping_address_name, "gstin")
+        != doc.billing_address_gstin,
+    )
```
Comment (ljain112) [inline india_compliance/gst_india/utils/e_invoice.py]: Can we directly use the response to update?

**Comment by ljain112** [comment]
@greptile review

File: india_compliance/gst_india/utils/e_waybill.py [hunk 14]
```diff
@@ -1274,19 +1275,36 @@ def get_data_with_irn(self):
...
+            "TransName": self.transaction_details.transporter_name,
+            "TransDocDt": self.transaction_details.lr_date,
+            "TransDocNo": self.transaction_details.lr_no,
+            "VehNo": self.transaction_details.vehicle_no,
+            "VehType": self.transaction_details.vehicle_type,
+        }
+        if (
+            is_e_waybill_changes_applicable(self.settings)
+            and self.transaction_details.transaction_type in SHIP_TO_TRANSACTION_TYPES
+            and not frappe.get_cached_value("e-Invoice Log", self.doc.irn, "is_generated_with_ship_to")
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/e_waybill.py]: simpler

File: india_compliance/gst_india/utils/e_invoice.py [hunk 15]
```diff
@@ -779,6 +786,23 @@ def set_transporter_details(self):
 
         return super().set_transporter_details()
 
+    def get_shipping_address(self):
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/e_invoice.py]: refactor

File: india_compliance/gst_india/utils/e_waybill.py [hunk 16]
```diff
@@ -1688,9 +1706,15 @@ def set_party_address_details(self):
             if side.gst_category == "SEZ":
                 side.state_number = 96
 
+        if has_different_to_address:
+            self.ship_to = self.get_address_details(address.ship_to)
+
+            # two addresses of the same party are a Regular transaction, since Ship To
```
Comment (vorasmit) [inline india_compliance/gst_india/utils/e_waybill.py]: refactor

**Comment by ljain112** [comment]
@greptile review

**Review by Anish59312 — COMMENTED** [review]
looks good to me

**Review by ljain112 — APPROVED** [review]
